src/include/mpir_hw_topo.h

```diff
@@ -0,0 +1,152 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef MPIR_HW_TOPO_H_INCLUDED
+#define MPIR_HW_TOPO_H_INCLUDED
+
+/*
+ * Node hardware object types
+ */
+typedef enum {
+    MPIR_NODE_HW_OBJ__MACHINE = 0,
+    MPIR_NODE_HW_OBJ__PACKAGE = 1,
+    MPIR_NODE_HW_OBJ__CORE = 2,
+    MPIR_NODE_HW_OBJ__PU = 3,
+    MPIR_NODE_HW_OBJ__L1CACHE = 4,
+    MPIR_NODE_HW_OBJ__L2CACHE = 5,
+    MPIR_NODE_HW_OBJ__L3CACHE = 6,
+    MPIR_NODE_HW_OBJ__L4CACHE = 7,
+    MPIR_NODE_HW_OBJ__L5CACHE = 8,
+    MPIR_NODE_HW_OBJ__L1ICACHE = 9,
+    MPIR_NODE_HW_OBJ__L2ICACHE = 10,
+    MPIR_NODE_HW_OBJ__L3ICACHE = 11,
+    MPIR_NODE_HW_OBJ__GROUP = 12,
+    MPIR_NODE_HW_OBJ__NUMANODE = 13,
+    MPIR_NODE_HW_OBJ__BRIDGE = 14,
+    MPIR_NODE_HW_OBJ__PCI_DEVICE = 15,
+    MPIR_NODE_HW_OBJ__OS_DEVICE = 16,
+    MPIR_NODE_HW_OBJ__MISC = 17,
+    MPIR_NODE_HW_OBJ__GPU_DEVICE = 18,
+    MPIR_NODE_HW_OBJ__NETWORK_DEVICE = 19,
+    MPIR_NODE_HW_OBJ__DDR = 20,
+    MPIR_NODE_HW_OBJ__HBM = 21,
+} MPIR_Node_hw_obj_type;
+
+/*
+ * Memory binding policies
+ */
+typedef enum {
+    MPIR_MEMBIND_POLICY__DEFAULT = 0,
+    MPIR_MEMBIND_POLICY__FIRSTTOUCH = 1,
+    MPIR_MEMBIND_POLICY__BIND = 2,
+    MPIR_MEMBIND_POLICY__INTERLEAVE = 3,
+    MPIR_MEMBIND_POLICY__NEXTTOUCH = 4,
+    MPIR_MEMBIND_POLICY__MIXED = -1
+} MPIR_Membind_policy;
```

Comment:
`MPIR_MEMBIND_POLICY__DEFAULT` should be outside the enum as a constant.  Also bonus points if it's a CVAR.

Reply:
Actually, I had a previous discussion point for this -- do you really mean `default` or `system-specified`?

Reply:
If the user says `default` there will be a system defined default not user defined. So it should not be a CVAR (unlike for the memory type). E.g., for linux it will be `firsttouch`, for solaris `bind`. Unless we want to change the behavior, in which case we have to expose this to be set by the user.

Reply:
I'd recommend calling it `SYSTEM`, instead of `DEFAULT`.

How does the user control which of these policies we'd use?  Isn't that a CVAR?  Should it be?

Reply:
These will be available to users through CVARs yes. Each buffer will have a `mem_type`, `mem_policy` and `mem_flags` (you can look at PR #3200 to see how they are defined for each buffer). This will be integrated in the other heterogeneous memory PR though. Later, if we find it useful, we can add more buffers.

Reply:
Since the memory binding pieces are not yet in `master`, it might be better to move this part to the other PR that introduces memory binding.

Reply:
Ok